sdk/core/azure-core/inc/azure/core/azure_assert.hpp

```diff
@@ -3,28 +3,55 @@
 
 /**
  * @file
- * @brief Provide assert macros to use as pre-conditions.
+ * @brief Provide assert macros to use with pre-conditions.
  *
- * @remark Asserts are turned ON even if `NDEBUG` is defined.
+ * @remark Asserts are turned ON when `NDEBUG` is NOT defined (for Debug build). For Release build,
+ * `std::abort()` is directly called if the condition is false, whitout calling assert().
  *
  */
 
 #pragma once
 
+#include "platform.hpp"
+
+#include <cstdlib>
+#include <string>
+
 #if defined(NDEBUG)
-#define AZURE_SDK_INTERNAL_IMPLEMENTATION_OVERRIDE_NDEBUG
-#undef NDEBUG
-#endif
+
+/*
+ * NDEBUG = defined = Build is on Release
+ * Define AZURE_ASSERT to call abort directly on exp == false
+ */
+
+#define AZURE_ASSERT(exp) \
+  if (!(exp)) \
+  { \
+    std::abort(); \
+  }
```

Comment:
This will do Bad Things™ if someone writes:
```c++
if (condition1)
  AZURE_ASSERT(condition2);
else
  do_stuff();
```
The user expects `do_stuff` to be called when `condition1` is `false`, but instead it will be called only when `condition1` and `condition2` are both `true` since the `else` binds to the `if` hidden inside `AZURE_ASSERT`. The idiomatic fix for this is to wrap the macro content in `do { /*...*/ } while(0)`:
```suggestion
#define AZURE_ASSERT(exp) \
  do { \
    if (!(exp)) \
    { \
      std::abort(); \
    } \
  while(0)
```